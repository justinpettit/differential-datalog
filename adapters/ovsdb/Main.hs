{-
Copyright (c) 2018 VMware, Inc.
SPDX-License-Identifier: MIT

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
-}

{-# LANGUAGE RecordWildCards #-}

import System.Environment
import Text.PrettyPrint
import System.Console.GetOpt
import Control.Exception
import Data.List
import Data.Maybe
import Data.List.Split
import Control.Monad
import qualified Data.Map as M

import Language.DifferentialDatalog.OVSDB.Compile

data TOption = OVSFile     String
             | OutputTable String
             | ROColumn    String
             | ProxyTable  String

options :: [OptDescr TOption]
options = [ Option ['f'] ["schema-file"]  (ReqArg OVSFile     "FILE")         "OVSDB schema file"
          , Option ['o'] ["output-table"] (ReqArg OutputTable "TABLE")        "mark TABLE as output"
          , Option []    ["ro"]           (ReqArg ROColumn    "TABLE.COLUMN") "mark COLUMN as read-only"
          , Option ['p'] ["gen-proxy"]    (ReqArg ProxyTable  "TABLE")        "generate output proxy table for TABLE"
          ]

data Config = Config { confOVSFile      :: FilePath
                     , confOutputTables :: [(String, [String])]
                     , confProxyTables  :: [String]
                     }

defaultConfig = Config { confOVSFile      = ""
                       , confOutputTables = []
                       , confProxyTables  = []
                       }


addOption :: Config -> TOption -> IO Config
addOption config (OVSFile f) = do
    when (confOVSFile config == "") $ errorWithoutStackTrace "Multiple input files specified"
    return config {confOVSFile = f}
addOption config (OutputTable t) = return config{ confOutputTables = nub ((t,[]) : confOutputTables config)}
addOption config (ProxyTable t) = return config{ confProxyTables = nub (t : confProxyTables config)}
addOption config (ROColumn c) = do
    case splitOn "." c of
         [table, col] -> do
            when (isNothing $ lookup table $ confOutputTables config)
                 $ errorWithoutStackTrace $ "Unknown output table name " ++ table
            let outtabs = map (\(t,ro) -> if t == table then (t, nub $ c:ro) else (t,ro))
                              $ confOutputTables config
            return $ config{confOutputTables = outtabs}
         _ -> errorWithoutStackTrace $ "Invalid column name " ++ c

validateConfig :: Config -> IO ()
validateConfig Config{..} = do
    when (confOVSFile == "") $ errorWithoutStackTrace "Input file not specified"

main = do
    args <- getArgs
    prog <- getProgName
    Config{..} <- case getOpt Permute options args of
                       (flags, [], []) -> do conf <- foldM addOption defaultConfig flags
                                             validateConfig conf
                                             return conf
                                          `catch`
                                          (\e -> do putStrLn $ usageInfo ("Usage: " ++ prog ++ " [OPTION...]") options
                                                    throw (e::SomeException))
                       _ -> errorWithoutStackTrace $ usageInfo ("Usage: " ++ prog ++ " [OPTION...]") options
    dlschema <- compileSchemaFile confOVSFile confOutputTables confProxyTables M.empty
    putStrLn $ render dlschema
    return ()
