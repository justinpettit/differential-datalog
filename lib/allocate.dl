
extern function allocate_u64(
    allocated: Set<bit<64>>,
    toallocate: Vec<'B>,
    max_val: bit<64>): Vec<('B, bit<64>)>
