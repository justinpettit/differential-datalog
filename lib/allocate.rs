pub fn allocate_allocate_u64<B: Ord+Clone>(allocated: &std_Set<u64>, toallocate: &std_Vec<B>, max_val: &u64) -> std_Vec<(B,u64)> {
    let first = *allocated.x.range(..).next_back().unwrap_or(&0);
    let mut last = first;
    let mut res = std_Vec::new();
    for b in toallocate.x.iter() {
        loop {
            last = if last + 1 <= *max_val {last + 1} else {1};
            if last == first {
                return res;
            };
            if allocated.x.contains(&last) { continue; };
            res.x.push(((*b).clone(), last));
            break;
        }
    };
    return res
}
