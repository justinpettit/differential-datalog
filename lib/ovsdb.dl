typedef integer      = bit<64>
typedef uuid         = bit<128>

typedef uuid_or_string_t = Either<uuid, string>

function extract_uuid (id: uuid_or_string_t): uuid = {
    match (id) {
        Left{uuid} -> uuid,
        _ -> 0
    }
}

extern function uuid2str(id: uuid): string

extern function set_extract_uuids(ids: Set<uuid_or_string_t>): Set<uuid>
extern function map_extract_val_uuids(ids: Map<'K, uuid_or_string_t>): Map<'K, uuid>

extern function group2vec_remove_sentinel(g: Group<uuid_or_string_t>): Vec<uuid_or_string_t>
extern function group2set_remove_sentinel(g: Group<uuid_or_string_t>): Set<uuid_or_string_t>
extern function group2map_remove_sentinel(g: Group<('K, uuid_or_string_t)>): Map<'K, uuid_or_string_t>

