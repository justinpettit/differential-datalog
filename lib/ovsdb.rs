use std::iter::FromIterator;
pub use ddlog_ovsdb_adapter::ovsdb_uuid2str;

pub fn ovsdb_map_extract_val_uuids<K: Val>(ids: &std_Map<K, ovsdb_uuid_or_string_t>) -> std_Map<K, ovsdb_uuid> {
    std_Map::from_iter(ids.x.iter().map(|(k,v)| (k.clone(), ovsdb_extract_uuid(v))))
}
pub fn ovsdb_set_extract_uuids(ids: &std_Set<ovsdb_uuid_or_string_t>) -> std_Set<ovsdb_uuid> {
    std_Set::from_iter(ids.x.iter().map(|x| ovsdb_extract_uuid(x)))
}


