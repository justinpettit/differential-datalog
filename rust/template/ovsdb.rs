//! OVSDB JSON interface to RunningProgram

use differential_datalog::program::*;
use std::ffi::{CStr};
use std::os::raw::{c_char, c_int};
use super::{Value, updcmd2upd};
use ddlog_ovsdb_adapter::*;
use std::sync;

/// Parse OVSDB JSON <table-updates> value into DDlog commands; apply commands to a DDlog program.
///
/// Must be called in the context of a transaction.
///
/// `prefix` contains is the prefix to be added to JSON table names, e.g, `OVN_Southbound_` or
/// `OVN_Northbound_` for OVN southbound and northbound database updates.
///
/// `updates` is the JSON string, e.g.,
/// ```
/// {"Logical_Switch":{"ffe8d84e-b4a0-419e-b865-19f151eed878":{"new":{"acls":["set",[]],"dns_records":["set",[]],"external_ids":["map",[]],"load_balancer":["set",[]],"name":"lsw0","other_config":["map",[]],"ports":["set",[]],"qos_rules":["set",[]]}}}}
/// ```
///
#[no_mangle]
pub extern "C" fn datalog_example_apply_ovsdb_updates(prog: *const sync::Mutex<RunningProgram<Value>>,
                                                      prefix: *const c_char,
                                                      updates: *const c_char) -> c_int
{
    let prog = unsafe {sync::Arc::from_raw(prog)};
    let res = apply_updates(&mut prog.lock().unwrap(), prefix, updates).map(|_|0).unwrap_or_else(|e|{
        eprintln!("datalog_example_apply_ovsdb_updates(): error: {}", e);
        -1
    });
    sync::Arc::into_raw(prog);
    res
}

fn apply_updates(prog: &mut RunningProgram<Value>, prefix: *const c_char, updates_str: *const c_char) -> Result<(), String>
{
    let prefix: &str = unsafe{ CStr::from_ptr(prefix) }.to_str()
        .map_err(|e|format!("invalid UTF8 string in prefix: {}", e))?;
    let updates_str: &str = unsafe{ CStr::from_ptr(updates_str) }.to_str()
        .map_err(|e|format!("invalid UTF8 string in prefix: {}", e))?;
    let commands = cmds_from_table_updates_str(prefix, updates_str)?;
    let updates: Result<Vec<Update<Value>>, String> = commands.iter().map(|c|updcmd2upd(c)).collect();
    prog.apply_updates(updates?)
}
