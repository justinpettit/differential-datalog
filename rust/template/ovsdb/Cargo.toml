[package]
name = "ddlog_ovsdb_adapter"
version = "0.1.0"

[dependencies.differential_datalog]
path = "../differential_datalog"

[dependencies]
serde_json="1.0"
num = "0.2"

[lib]
name = "ddlog_ovsdb_adapter"
path = "./lib.rs"
crate-type = ["staticlib", "rlib"]
