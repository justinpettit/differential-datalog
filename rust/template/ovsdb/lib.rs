//! Parse OVSDB database update messages and convert them into DDlog table update commands

extern crate differential_datalog;
extern crate serde_json;
extern crate num;

use differential_datalog::record::*;
use serde_json::Value;
use serde_json::map::Map;
use num::BigInt;

/*
pub enum UpdCmd {
    Insert (String, Record),
    Delete (String, Record),
    DeleteKey(String, Record)
}
*/

fn val_into_vec(v: Value) -> Vec<Value> {
    match v {
        Value::Array(a) => a,
        _ => panic!("not an array value")
    }
}

fn complain<X>(src: &Value, category: &str) -> Result<X, String>{
    Err(format!("JSON value {} is not a valid {}", src, category))
}

fn lookup<'a>(m: &'a mut Map<String, Value>, field: &str) -> Result<Value, String> {
    m.remove(field).ok_or(format!("JSON map does not contain \"{}\" field: {:?}", field, m))
}

/*
 {
    "op": "update",
    "row": {
      "bridges": [
        "named-uuid",
        "row9b73f7bc_2a67_48df_88b0_fabc19686920"]},
    "table": "Open_vSwitch",
    "where": [
      [
        "_uuid",
        "==",
        [
          "uuid",
          "cf9dca71-2495-4e61-9fe2-8ea4fb2859cd"]]]
 }
*/
pub fn cmd_from_val(src: Value) -> Result<UpdCmd, String> {
    match src {
        Value::Object(mut m) => {
            let op: String = serde_json::from_value(lookup(&mut m, "op")?.clone()).map_err(|e|e.to_string())?;
            let table: String = serde_json::from_value(lookup(&mut m,"table")?.clone()).map_err(|e|e.to_string())?;
            match op.as_ref() {
                "update" => { 
                    let rec = row_from_val(m, &table)?;
                    Ok(UpdCmd::Insert(table, rec))
                },
                _ => Err(format!("unsupported operation {} in {:?}", op, m))
            }
        },
        _ => complain(&src, "command")
    }
}

fn row_from_val(mut m: Map<String, Value>, table: &str) -> Result<Record, String> {
    // extract _uuid from "where"-condition of the form
    // "where": [[ "_uuid", "==", ["uuid", "cf9dca71-2495-4e61-9fe2-8ea4fb2859cd"]]]
    let uuid = match lookup(&mut m, "where")? {
        Value::Array(mut v) => {
            let valid_pattern = match v.as_slice() {
                [Value::String(field),Value::String(op), Value::Array(_)] => {
                    field == "_uuid" && op == "=="
                },
                _ => false
            };
            if valid_pattern {
                record_from_array(val_into_vec(v.remove(2)))
            } else {
                Err(format!("unsupported \"where\" condition: {:?}", v))
            }
        },
        _ => Err(format!("\"where\" condition is not an array in {:?}", m))
    }?;

    // parse row
    let mut fields = match lookup(&mut m, "row")? {
        Value::Object(obj) => row_from_obj(obj),
        _ => Err(format!("\"row\" is not an object in {:?}", m))
    }?;
    fields.push(("_uuid".to_owned(), uuid));
    Ok(Record::NamedStruct(table.to_string(), fields))
}

/*
    ["uuid", "cf9dca71-2495-4e61-9fe2-8ea4fb2859cd"]
 */
fn record_from_array(mut src: Vec<Value>) -> Result<Record,String> {
    if src.len() != 2 {
        return Err(format!("record array is not of length 2: {:?}", src));
    };
    let val = src.remove(1);
    let field = src.remove(0);
    match field {
        Value::String(field) => {
            match (field.as_str(), val) {
                ("uuid", Value::String(uuid)) => {
                    Ok(Record::Int(BigInt::parse_bytes(uuid.as_bytes(), 16)
                                   .ok_or(format!("invalid uuid string \"{}\"", uuid))?))
                },
                ("named-uuid", Value::String(uuid_name)) => {
                    Ok(Record::String(uuid_name))
                },
                ("set", Value::Array(atoms)) => {
                    let elems: Result<Vec<Record>, String> = atoms.into_iter().map(|a|record_from_val(a)).collect();
                    Ok(Record::Array(elems?))
                },
                ("map", Value::Array(pairs)) => {
                    let elems: Result<Vec<(Record, Record)>, String> = pairs.into_iter().map(|p|pair_from_val(p)).collect();
                    Ok(Record::Array(elems?.into_iter().map(|(k,v)|Record::Tuple(vec![k,v])).collect()))
                },
                _ => Err(format!("unexpected array value type \"{}\"", field))
            }
            
        },
        _ => Err(format!("invalid array value: {:?}", src))
    }
}

fn pair_from_val(pair: Value) -> Result<(Record, Record), String> {
    match pair {
        Value::Array(mut v) => {
            if v.len() != 2 {
                return Err(format!("key-value pair must be of length 2: {:?}",v))
            };
            let val = v.remove(1);
            let key = v.remove(0);
            Ok((record_from_val(key)?, record_from_val(val)?))
        },
        _ => Err(format!("invalid key-value pair {}", pair))
    }
}

/*
  {
    "interfaces": [ "named-uuid", "rowe82c26a4_bbf4_4e84_87fd_2107d5998a12"],
    "name": "br0"
  }
*/
fn row_from_obj(m: Map<String, Value>) -> Result<Vec<(String, Record)>, String> {
    m.into_iter().map(|(field, val)|Ok((field, record_from_val(val)?))).collect()
}

fn record_from_val(v: Value) -> Result<Record, String> {
    match v {
        // <string>
        Value::String(s) => Ok(Record::String(s)),
        // <number>
        Value::Number(ref n) if n.is_u64() => Ok(Record::Int(BigInt::from(n.as_u64().unwrap()))),
        // <boolean>
        Value::Bool(b) => Ok(Record::Bool(b)),
        // <uuid>, <named-uuid>, <set>, <map>
        Value::Array(v) => record_from_array(v),
        _ => Err(format!("unexpected value {}", v))
    } 
}
