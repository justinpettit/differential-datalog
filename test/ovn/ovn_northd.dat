start;

insert OVN_Northbound.Meter_Band (
    ._uuid = 0x06a2cf7c_7ea1_4ca2_a0a2_6176fd1c6d95,
    .action = "drop",
    .rate = 1000,
    .burst_size = 0,
    .external_ids = []
),

insert OVN_Northbound.Meter (
    ._uuid = 0x06a2cf7c_7ea1_4ca2_a0a2_6176fd1c6d92,
    .name = "meter1",
    .unit = "kbps",
    .bands = [0x06a2cf7c_7ea1_4ca2_a0a2_6176fd1c6d95],
    .external_ids = []
),

insert OVN_Northbound.Logical_Switch (
    ._uuid = 0x06a2cf7c_7ea1_4ca2_a0a2_6176fd1c6d90,
    .name = "ls1",
    .ports = [],
    .acls = [],
    .qos_rules = [],
    .load_balancer = [],
    .dns_records = [],
    .other_config = [],
    .external_ids = [("neutron:network_name","neutron_network_1")]
),

insert OVN_Southbound.Datapath_Binding(
    ._uuid = 0x06a2cf7c_7ea1_4ca2_a0a2_6176fd1c6d11,
    .tunnel_key = 1,
    .external_ids = [("logical-switch", "06a2cf7c-7ea1-4ca2-a0a2_6176fd1c6d12")]
);

commit;
dump;

echo Clearing Datapath_Binding;

start;

clear OVN_Southbound.Datapath_Binding;

commit;

dump OVN_Southbound.DeltaPlus_Datapath_Binding;
dump OVN_Southbound.DeltaMinus_Datapath_Binding;
