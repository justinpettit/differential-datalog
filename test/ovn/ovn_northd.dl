import OVN_Northbound as nb
import OVN_Southbound as sb
import ovsdb
import allocate

/* Meter_Band table */
for (mb in nb.Meter_Band) {
    sb.Meter_Band(.uuid_name = uuid2str(mb._uuid),
                  .action = mb.action,
                  .rate = mb.rate,
                  .burst_size = mb.burst_size)
}


/* Meter table */
sb.Meter(name, unit, str_bands) :-
    nb.Meter(_, name, unit, bands, _),
    var uuid_band = FlatMap(bands),
    var str_band = uuid2str(uuid_band),
    Aggregate((name, unit), str_bands = group2set(str_band)).


/* Datapath_Binding table */
sb.Datapath_Binding(uuid_str, tunkey, external_ids) :-
    nb.Logical_Switch(._uuid = uuid, .name = name, .external_ids = ids),
    var uuid_str = uuid2str(uuid),
    TunKeyAllocation(uuid_str, tunkey),
    var external_ids = {
        var eids: Map<string, string> = map_empty();
        map_insert(eids, "logical-switch", uuid_str);
        map_insert(eids, "name", name);
        match (map_get(ids, "neutron:network-name")) {
            None -> (),
            Some{nnn} -> map_insert(eids, "name2", nnn)
        };
        eids
    }.

/* tunnel key allocation */
output relation AllocatedTunKeys(logical_id: string, tunkey: integer)

// compute a set of already allocated keys from the Realized_Datapath_Binding table.
AllocatedTunKeys(ls_id, tunkey) :-
    sb.Realized_Datapath_Binding(.tunnel_key = tunkey,
                                 .external_ids = eids),
    Some{var ls_id} = map_get(eids, "logical-switch").

AllocatedTunKeys(lr_id, tunkey) :-
    sb.Realized_Datapath_Binding(.tunnel_key = tunkey,
                                 .external_ids = eids),
    Some{var lr_id} = map_get(eids, "logical-router").

output relation AggAllocatedTunKeys(allkeys: Set<integer>)

AggAllocatedTunKeys(allkeys) :-
    AllocatedTunKeys(_, tunkey),
    Aggregate((), allkeys = group2set(tunkey)).

// compute a vector of logical switch and logical router ids that
// have not been allocated keys yet

output relation NotYetAllocatedTunKeys(logical_id: string)

NotYetAllocatedTunKeys(ls_id) :-
    nb.Logical_Switch(._uuid = uuid),
    var ls_id = uuid2str(uuid),
    not AllocatedTunKeys(ls_id, _).

NotYetAllocatedTunKeys(lr_id) :-
    nb.Logical_Router(._uuid = uuid),
    var lr_id = uuid2str(uuid),
    not AllocatedTunKeys(lr_id, _).

output relation AggNotYetAllocatedTunKeys(all_logical_ids: Vec<string>)

AggNotYetAllocatedTunKeys(all_logical_ids) :-
    NotYetAllocatedTunKeys(lid),
    Aggregate((), all_logical_ids = group2vec(lid)).

// Perform the allocation
output relation TunKeyAllocation(logical_id: string, tunkey: integer)

// transfer existing allocations as is
TunKeyAllocation(lid, tunkey) :- AllocatedTunKeys(lid, tunkey).

// Case 1: AggAllocatedTunKeys relation is not empty (i.e., contains
// a single record that stores a set of allocated keys)
TunKeyAllocation(lid, tunkey) :-
    AggAllocatedTunKeys(allocated),
    AggNotYetAllocatedTunKeys(unallocated),
    var allocation = FlatMap(allocate_u64(allocated, unallocated, 64'hffff)),
    (var lid, var tunkey) = allocation.

// Case 2: AggAllocatedTunKeys relation is empty 
TunKeyAllocation(lid, tunkey) :-
    AggNotYetAllocatedTunKeys(unallocated),
    not AggAllocatedTunKeys(_),
    var allocation = FlatMap(allocate_u64(set_empty(), unallocated, 64'hffff)),
    (var lid, var tunkey) = allocation.
