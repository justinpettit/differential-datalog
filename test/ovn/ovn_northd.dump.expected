AggAllocatedTunKeys:
AggAllocatedTunKeys{std_Set { x: {1} }}

AggNotYetAllocatedTunKeys:
AggNotYetAllocatedTunKeys{std_Vec { x: ["906d1cfd-7661-a2a0-a24c-a17e7ccfa206"] }}

AllocatedTunKeys:
AllocatedTunKeys{"06a2cf7c-7ea1-4ca2-a0a2_6176fd1c6d12",1}

NotYetAllocatedTunKeys:
NotYetAllocatedTunKeys{"906d1cfd-7661-a2a0-a24c-a17e7ccfa206"}

OVN_Southbound_DeltaMinus_Datapath_Binding:
OVN_Southbound.DeltaMinus_Datapath_Binding{8820728388062355172047994593891609873}

OVN_Southbound_DeltaPlus_Datapath_Binding:
OVN_Southbound.DeltaPlus_Datapath_Binding{"906d1cfd-7661-a2a0-a24c-a17e7ccfa206",2,std_Map { x: {"logical-switch": "906d1cfd-7661-a2a0-a24c-a17e7ccfa206", "name": "ls1", "name2": "neutron_network_1"} }}

OVN_Southbound_DeltaPlus_Meter:
OVN_Southbound.DeltaPlus_Meter{"meter1","kbps",std_Set { x: {std.Right{"956d1cfd-7661-a2a0-a24c-a17e7ccfa206"}} }}

OVN_Southbound_DeltaPlus_Meter_Band:
OVN_Southbound.DeltaPlus_Meter_Band{"956d1cfd-7661-a2a0-a24c-a17e7ccfa206","drop",1000,0}

TunKeyAllocation:
TunKeyAllocation{"06a2cf7c-7ea1-4ca2-a0a2_6176fd1c6d12",1}
TunKeyAllocation{"906d1cfd-7661-a2a0-a24c-a17e7ccfa206",2}

